python/pyarrow/array.pyx

```diff
@@ -402,6 +402,31 @@ cdef class BinaryArray(Array):
 
 cdef class DictionaryArray(Array):
 
+    cdef getitem(self, int64_t i):
+        cdef Array dictionary = self.dictionary
+        cdef int64_t index = self.indices[i].as_py()
+        return scalar.box_scalar(dictionary.type, dictionary.sp_array, index)
+
+    property dictionary:
```

Comment:
This is the @cache_readonly pattern used elsewhere. Thread safety doesn't buy us anything, and I don't see a reason to create two objects that may never be used.